### PR TITLE
Add evaluatesourcenodes to evaluation 

### PR DESCRIPTION
**Pull request by ravi03071991**
Hi,

QueryResponseEvaluator now includes the feature 'evaluatesourcenodes' which enables identification of the source nodes containing the answer.

Thank you.

### CODE REVIEW

File: gpt_index/evaluation/base.py [hunk 1]
```diff
@@ -218,3 +218,51 @@ def evaluate(self, query: str, response: Response) -> str:
...
+            Yes -> If answer, context information are matching \
+                    or If Query, answer and context information are matching \
+                        for a source node.
+            No -> If answer, context information are not matching \
+                    or If Query, answer and context information are not matching \
+                        for a source node.
+        """
+        answer = str(response)
+
+        context = self.get_context(response)
```
Comment (jerryjliu) [inline gpt_index/evaluation/base.py:240]: nit: rename to context_list (since it's a list of sources)

File: gpt_index/evaluation/base.py [hunk 2]
```diff
@@ -218,3 +218,51 @@ def evaluate(self, query: str, response: Response) -> str:
...
+
+        response_source_node = []
```
Comment (jerryjliu) [inline gpt_index/evaluation/base.py:242]: i'd rename this to response_texts

File: gpt_index/evaluation/base.py [hunk 3]
```diff
@@ -218,3 +218,51 @@ def evaluate(self, query: str, response: Response) -> str:
...
+            )
+
+            query_response = f"Question: {query}\nResponse: {answer}"
+
+            response_obj = index.query(
+                query_response,
+                text_qa_template=QUERY_RESPONSE_EVAL_PROMPT_TMPL,
+                refine_template=QUERY_RESPONSE_REFINE_PROMPT_TMPL,
+            )
+            response_txt = str(response_obj)
```
Comment (jerryjliu) [inline gpt_index/evaluation/base.py:264]: is there a chance response_txt doesn't conform to the right response? 

maybe we should do some basic parsing like 
```python
if "yes" in raw_response_txt.lower():
  response_txt = "YES"
elif "no" in raw_response_txt.lower():
  response_txt = "NO"
else:
   raise ValueError("The response is invalid")
```

**Review by jerryjliu — COMMENTED** [review]
high-level looks pretty good! some smaller comments

File: gpt_index/evaluation/base.py (hunk 3 above)
Comment (ravi03071991) [reply gpt_index/evaluation/base.py:264]: Did not find such case in my experiments but will add as a fall back.

**Review by jerryjliu — APPROVED** [review]
thanks!

**Pull request merged**